src/comm/TcpAcceptor.cc

```diff
@@ -390,10 +390,17 @@ Comm::TcpAcceptor::oldAccept(Comm::ConnectionPointer &details)
     details->local = *gai;
     Ip::Address::FreeAddr(gai);
 
-    // Perform NAT or TPROXY operations to retrieve the real client/dest IP addresses
-    if (conn->flags&(COMM_TRANSPARENT|COMM_INTERCEPTION) && !Ip::Interceptor.Lookup(details, conn)) {
-        debugs(50, DBG_IMPORTANT, "ERROR: NAT/TPROXY lookup failed to locate original IPs on " << details);
-        return Comm::NOMESSAGE;
+    if (conn->flags & COMM_TRANSPARENT) { // the real client/dest IP address must be already available via getsockname()
+        // XXX: a reconfiguration may 'stop' transparency in Interceptor (failing this check).
+        // We need to keep consistency between Interceptor and Comm flags.
+        assert(Ip::Interceptor.TransparentActive());
+        details->flags |= COMM_TRANSPARENT;
+    } else if (conn->flags & COMM_INTERCEPTION) { // request the real client/dest IP address from NAT
+        details->flags |= COMM_INTERCEPTION;
+        if (!Ip::Interceptor.LookupNat(*details)) {
+            debugs(50, DBG_IMPORTANT, "ERROR: NAT lookup failed to locate original IPs on " << details);
+            return Comm::NOMESSAGE;
```

Comment:
IIRC we have had already a discussion about this in PR837 (9fd3e68c) and rejected this idea.

Reply:
That PR was for a specific bug and my objection was overruled on grounds the bug fix was important enough to ignore this bug. This PR is a refactor and should do the right thing instead of adding more bugs.

The NOMESSAGE return value is/was designed specifically for the case where a worker does **NOT** receive a connection from accept(2). The code handling that output assumes the ```Comm::Connection::fd``` value is ```-1``` (aka closed) when ```Comm::TcpAcceptor``` frees it implicitly.

The situation with NAT (and now TPROXY as well) is that an **error** happened after control of the connection was taken on by TcpAcceptor. As I said the connection is already fully open and potentially has traffic underway.

Correct usage of a Comm::Connection when storing a TCP connection is to use ```close()``` on errors.

FYI, The class destructors RAII closure cannot be relied on because Comm::Connection stores FD for protocols (eg SNMP, UDP, ICP, HTCP) that do not have a requirement for ```close()``` to happen. We are **lucky** those components of Squid handle their Comm::Connection in a synchronous callback sequence right now. As Squid code is converted to AsyncJob's that luck is going away.

Reply:
> Alex: We will address it in a dedicated PR which I volunteer to promptly post.

Done: I have a Draft PR with a sketch of a clean solution for controlling the accepted socket lifetime in the presence of errors: https://github.com/measurement-factory/squid/pull/161

The draft PR may need some work, but I believe it is on the right path. I drafted that PR on Factory side so that GitHub shows the right diff against _this_ PR. When this PR is merged, we can move that draft PR here, to this official repository, so that it can be properly discussed and advanced. For now, the draft can be used as an illustration of how the problem raised in this change request may be solved and to confirm that the problem (and the solution) are outside of (or orthogonal to) this PR scope.

Reply:
Very well. I have looked at that Draft and can agree to its design being the resolution of this issue, though there are some things I would like to see different we shall discuss there.

Resolving this request.

Reply:
Thank you. That internal draft PR has been updated and posted as official PR #1103.